### PR TITLE
Implement Entry API, Rebased #37

### PR DESCRIPTION
**Pull request by reem**

### CODE REVIEW

**Comment by reem** [comment]
The code looks good to me, but probably someone else should sign off, @bluss @apasel422?

File: src/lib.rs [hunk 1]
```diff
@@ -164,6 +193,79 @@ impl<K: Hash + Eq, V, S: BuildHasher> LinkedHashMap<K, V, S> {
...
+            });
+        }
+
+        Entry::Vacant(VacantEntry {
+            key: k,
+            map: self,
+        })
+    }
+
+    /// Returns an iterator visiting all entries in order insertion.
```
Comment (apasel422) [inline src/lib.rs]: Should this be "in insertion order"?

File: src/lib.rs [hunk 2]
```diff
@@ -877,6 +993,163 @@ impl<'a, K: Hash + Eq, V, S: BuildHasher> IntoIterator for &'a mut LinkedHashMap
...
+}
+
+/// A view into a single occupied location in a LinkedHashMap.
+pub struct OccupiedEntry<'a, K: 'a, V: 'a, S: 'a = hash_map::RandomState> {
+    entry: *mut LinkedHashMapEntry<K, V>,
+    map: *mut LinkedHashMap<K, V, S>,
+    marker: marker::PhantomData<&'a K>,
+}
+
+/// A view into a single empty location in a HashMap.
```
Comment (apasel422) [inline src/lib.rs]: s/HashMap/LinkedHashMap/

**Comment by FlashCat** [comment]
:umbrella: The latest upstream changes (presumably #44) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by reem** [comment]
@apasel422 rebased and updated for comments.

File: src/lib.rs [hunk 3]
```diff
@@ -592,27 +688,10 @@ impl<'a, K, V, S, Q: ?Sized> IndexMut<&'a Q> for LinkedHashMap<K, V, S>
...
-            (*node).next = (*self.head).next;
-            (*node).prev = self.head;
-            (*self.head).next = node;
-            (*(*node).next).prev = node;
-        }
-    }
-}
-
 #[cfg(not(feature = "nightly"))]
+// FIXME: `HashMap` doesn't expose its hash state, so we cannot clone fully parameterized
```
Comment (apasel422) [inline src/lib.rs]: I think this comment was mistakenly reintroduced?

**Comment by apasel422** [comment]
Ping.

**Comment by apasel422** [comment]
@reem Any update on this?

**Comment by zonyitoo** [comment]
Any update on this?

**Comment by FlashCat** [comment]
:umbrella: The latest upstream changes (presumably #54) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by dtolnay** [comment]
@reem @apasel422 is there anything I can do to help move this along?

**Comment by apasel422** [comment]
Perhaps rebase it and submit a new PR?

**Comment by dtolnay** [comment]
Rebased in #76.

**Pull request merged**